tests/unittests/dhcp/kea_metrics_test.py

```diff
@@ -0,0 +1,433 @@
+    {"result": 4, "text": "b", "arguments": {"arg1": "val1"}, "service": "d"},
+    {"text": "b", "arguments": {"arg1": "val1"}, "service": "d"
+    ]'''
+
+
+@pytest.fixture
+def large_response():
+    return '''
+
+    '''
+
+
+def send_dummy_query(command="command"):
+    return send_query(
+        query=KeaQuery(command, [], {}),
+        address="192.0.2.2",
+        port=80,
+    )
+
+
+################################################################################
+# Testing the list[KeaResponse] returned by send_query()                       #
+################################################################################
+
+
+def test_success_responses_does_succeed(success_response, enqueue_post_response):
+    enqueue_post_response("command", success_response)
+    responses = send_dummy_query("command")
+    assert len(responses) == 4
+    for response in responses:
+        assert response.success
+        assert isinstance(response.text, str)
+        assert isinstance(response.arguments, dict)
+        assert isinstance(response.service, str)
+
+
+def test_error_responses_does_not_succeed(error_response, enqueue_post_response):
+    enqueue_post_response("command", error_response)
+    responses = send_dummy_query("command")
+    assert len(responses) == 5
+    for response in responses:
+        assert not response.success
+        assert isinstance(response.text, str)
+        assert isinstance(response.arguments, dict)
+        assert isinstance(response.service, str)
+
+
+def test_invalid_json_responses_raises_jsonerror(
+    invalid_json_response, enqueue_post_response
+):
+    enqueue_post_response("command", invalid_json_response)
+    with pytest.raises(JSONDecodeError):
+        responses = send_dummy_query("command")
+
+
+################################################################################
+# Testing KeaDhcpSubnet and KeaDhcpConfig instantiation from json              #
+################################################################################
+
+
+def test_correct_subnet_from_dhcp4_config_json(dhcp4_config):
+    j = json.loads(dhcp4_config)
+    subnet = KeaDhcpSubnet.from_json(j["Dhcp4"]["subnet4"][0])
+    assert subnet.id == 1
+    assert subnet.prefix == IP("192.0.0.0/8")
+    assert len(subnet.pools) == 2
+    assert subnet.pools[0] == (IP("192.1.0.1"), IP("192.1.0.200"))
+    assert subnet.pools[1] == (IP("192.3.0.1"), IP("192.3.0.200"))
+
+
+def test_correct_config_from_dhcp4_config_json(dhcp4_config):
+    j = json.loads(dhcp4_config)
+    config = KeaDhcpConfig.from_json(j)
+    assert len(config.subnets) == 1
+    subnet = config.subnets[0]
+    assert subnet.id == 1
+    assert subnet.prefix == IP("192.0.0.0/8")
+    assert len(subnet.pools) == 2
+    assert subnet.pools[0] == (IP("192.1.0.1"), IP("192.1.0.200"))
+    assert subnet.pools[1] == (IP("192.3.0.1"), IP("192.3.0.200"))
+    assert config.dhcp_version == 4
+    assert config.config_hash is None
```

Comment:
Not entirely sure what these tests are really testing for, so I'm having a hard time suggesting alternate names here:

I generally prefer that tests are named very explicitly so that when I'm looking at a test report and see a failing test, I won't have to dig deeply into the test code itself to figure out what requirement it was that actually failed.

Names along the lines of "when x happens y should do z" (or, if grouping tests in classes to test individual functions or classes in multiple ways, "when x happens it should do z")

Reply:
Thank you for helpful input. Make sure to continue pointing out any conventions, guidelines or modules I've missed out on!

Regarding naming of test functions. I agree that the function names are unclear. I've not followed any convention, but for those places where I'm testing a single function, the test function's name usually has been some combination of `fail|correct|invalid` , `function name`, and  `function input`. I'll begin to name my test functions in such a way that the name becomes a summary of the test's intention. 

Regarding the two odd tests. `test_correct_config_from_dhcp4_config_json()`  checks that we've correctly processed the config json of some Kea dhcp server into a dataclass that is more ergonomic to work with, i.e. that we've coalesced *all* subnets into a single python list and processed *strings* that represented the pool ranges of subnets correctly into tuples (this last processing step may not be needed by Nav; we only seem to care about subnets and not pools). The following name should make the intent clearer:
```diff
- def test_correct_config_from_dhcp4_config_json(dhcp4_config):
+ def test_dhcp4_config_json_is_correctly_processed_into_config_dataclass(dhcp4_config):
```

`test_correct_subnet_from_dhcp4_config_json()` actually just tests a subset of the above test, so in hindsight I see that it's not needed.

Regarding Pydantic. There's definetively a use-case for Pydantic here. Validating takes up big portions of the code and is not fun to read.. Some manual processing must alas be done so long as we use `config-get` instead of the more well tailored but not universally available `subnet4-list` command of the API to fetch subnet information.